### PR TITLE
jobs demo: add jobs demo

### PR DESCRIPTION
**Pull request by nateglims**
* Add demo that exercises the jobs library.
* Add demo to demo CMakeLists.txt.

*Issue #, if available:*

*Description of changes:*

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: demos/source/aws_iot_demo_jobs.c [hunk 1]
```diff
@@ -0,0 +1,528 @@
...
+            {
+                IotLogInfo( "Failed to execute print." );
+                result = AWS_IOT_JOB_STATE_FAILED;
+            }
+            break;
+        case 'w':
+            result = AWS_IOT_JOB_STATE_SUCCEEDED;
+            IotClock_SleepMs( 10000 );
+            /*
+	    if ( _getMessage( &pMessage, &messageLength ) )
```
Comment (nateglims) [inline demos/source/aws_iot_demo_jobs.c]: I'm not sure if I should pull in `strtol` or similar for this.

File: demos/source/aws_iot_demo_jobs.c [hunk 2]
```diff
@@ -0,0 +1,533 @@
...
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+ * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+ * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+ * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+/**
+ * @file aws_iot_demo_jobs.c
+ * @brief Demonstrates use of the AWS Jobs library.
```
Comment (gordonwang0) [inline demos/source/aws_iot_demo_jobs.c]: Minor: should be "AWS IoT Jobs library"

File: demos/source/aws_iot_demo_jobs.c [hunk 3]
```diff
@@ -0,0 +1,533 @@
...
+
+/* MQTT Include. The underlying protocol for this demo. */
+#include "iot_mqtt.h"
+
+/* JSON Utilies for pulling out values. */
+#include "iot_json_utils.h"
+
+/* Jobs Library. */
+#include "aws_iot_jobs.h"
+#include "types/aws_iot_jobs_types.h"
```
Comment (gordonwang0) [inline demos/source/aws_iot_demo_jobs.c]: You don't need to include the types header, the library header already includes it.

File: demos/source/aws_iot_demo_jobs.c [hunk 4]
```diff
@@ -0,0 +1,533 @@
...
+/**
+ * @brief The Jobs Library MQTT timeout.
+ * @TODO: Figure out how this interacts with the MQTT library timeout.
+ */
+#define AWS_IOT_DEMO_JOBS_MQTT_TIMEOUT ( 1000U )
+
+/**
+ * @brief The MQTT Library timeout.
+ */
+#define TIMEOUT_MS ( 5000u )
```
Comment (gordonwang0) [inline demos/source/aws_iot_demo_jobs.c]: Should consolidate these two timeout values to 1 value

File: demos/source/aws_iot_demo_jobs.c [hunk 5]
```diff
@@ -0,0 +1,533 @@
...
+
+/* Jobs Library. */
+#include "aws_iot_jobs.h"
+#include "types/aws_iot_jobs_types.h"
+
+/**
+ * @brief The Jobs Library MQTT timeout.
+ * @TODO: Figure out how this interacts with the MQTT library timeout.
+ */
+#define AWS_IOT_DEMO_JOBS_MQTT_TIMEOUT ( 1000U )
```
Comment (gordonwang0) [inline demos/source/aws_iot_demo_jobs.c]: This is probably too short.

File: demos/source/aws_iot_demo_jobs.c [hunk 6]
```diff
@@ -0,0 +1,533 @@
...
+ */
+static char _jobDoc[ JOBS_DEMO_MAX_JOB_DOC_LENGTH + 1 ] = {0};
+static size_t _jobDocLength = 0;
+
+/**
+ * @brief Execute the Jobs Demo.
+ *
+ * This is the top level function called by the IOT Demo harness.
+ */
+int32_t RunJobsDemo( bool awsIotMqttMode,
```
Comment (gordonwang0) [inline demos/source/aws_iot_demo_jobs.c]: Return value should just be `int`, for compatibility with `EXIT_SUCCESS` and `EXIT_FAILURE`

File: demos/source/aws_iot_demo_jobs.c [hunk 7]
```diff
@@ -0,0 +1,533 @@
...
+    mqttInitStatus = IotMqtt_Init();
+
+    if( mqttInitStatus != IOT_MQTT_SUCCESS )
+    {
+        status = EXIT_FAILURE;
+    }
+
+    if( status == EXIT_SUCCESS )
+    {
+        jobsInitStatus = AwsIotJobs_Init( AWS_IOT_DEMO_JOBS_MQTT_TIMEOUT );
```
Comment (gordonwang0) [inline demos/source/aws_iot_demo_jobs.c]: It's sufficient just to pass 0 ("use default").

File: demos/source/aws_iot_demo_jobs.c [hunk 8]
```diff
@@ -0,0 +1,533 @@
...
+static void _jobsCallback( void * param1,
+                           AwsIotJobsCallbackParam_t * pCallbackInfo )
+{
+    bool success = false;
+    const char * pJsonValue = NULL;
+    size_t jsonValueLength = 0;
+    IotSemaphore_t * pWaitSem = ( IotSemaphore_t * ) param1;
+
+    IotLogInfo( "Job Doc Len : %d", pCallbackInfo->u.callback.documentLength );
+    IotLogInfo( "Job Document: %s", pCallbackInfo->u.callback.pDocument );
```
Comment (gordonwang0) [inline demos/source/aws_iot_demo_jobs.c]: Job document is not guaranteed to be NULL-terminated.  Must be printed with the length specifier.

Also, printing the length is probably not necessary.

File: demos/source/aws_iot_demo_jobs.c [hunk 9]
```diff
@@ -0,0 +1,533 @@
...
+                                    &jsonValueLength );
+    if( success )
+    {
+        IotLogInfo( "New Job: %.*s", jsonValueLength, pJsonValue );
+        memcpy( _pJobId, pJsonValue, jsonValueLength );
+        _JobIdLength = jsonValueLength;
+    }
+    else
+    {
+        IotLogError( "Failed to parse JobID from Notify Next." );
```
Comment (gordonwang0) [inline demos/source/aws_iot_demo_jobs.c]: This function should exit (ignore Job) if it can't find the key it wants.  No point in continuing to parse.

File: demos/source/aws_iot_demo_jobs.c [hunk 10]
```diff
@@ -0,0 +1,533 @@
...
+                                    pCallbackInfo->u.callback.documentLength,
+                                    JOB_DOC_KEY,
+                                    JOB_DOC_KEY_LENGTH,
+                                    &pJsonValue,
+                                    &jsonValueLength );
+
+    if( success )
+    {
+        IotLogInfo( "New Job Doc: %.*s", jsonValueLength, pJsonValue );
+        memcpy( _jobDoc, pJsonValue, jsonValueLength );
```
Comment (gordonwang0) [inline demos/source/aws_iot_demo_jobs.c]: If jsonValueLength is longer than _jobDoc's length, then _jobDoc will contain a partial and probably invalid document.  This needs to be handled somehow.

File: demos/source/aws_iot_demo_jobs.c [hunk 11]
```diff
@@ -0,0 +1,533 @@
...
+            IotLogError( "Failed to find %s Key.", JOB_ACTION_KEY );
+        }
+    }
+
+    /* Process the action. */
+    if( success )
+    {
+        /* We are only going to look at the first char after the '"'.
+	   This is just to simplify the demo. */
+        switch( pAction[ 1 ] )
```
Comment (gordonwang0) [inline demos/source/aws_iot_demo_jobs.c]: Need to assume the action can be anything, this should be a string compare.

File: demos/source/aws_iot_demo_jobs.c [hunk 12]
```diff
@@ -0,0 +1,533 @@
...
+		result = AWS_IOT_JOB_STATE_SUCCEEDED;
+	    }
+            else
+            {
+                IotLogInfo( "Failed to execute wait." );
+                result = AWS_IOT_JOB_STATE_FAILED;
+            }
+	    */
+            break;
+        default:
```
Comment (gordonwang0) [inline demos/source/aws_iot_demo_jobs.c]: Maybe add a "publish" action.  For example:

`{"action":"publish","message":"Hello world!"}` sends "Hello world!" on the MQTT connection.

Comment (nateglims) [reply demos/source/aws_iot_demo_jobs.c]: What topic?

Comment (gordonwang0) [reply demos/source/aws_iot_demo_jobs.c]: Have the topic in the Job document.

```{"action":"publish","message":"Hello world!","topic":"demo/jobs"}```

File: demos/source/aws_iot_demo_jobs.c [hunk 13]
```diff
@@ -0,0 +1,591 @@
...
+
+/**
+ * @brief Initialize The Jobs Demo.
+ *
+ * Initialize MQTT module then the Jobs module. If the Jobs module fails
+ * to initialize, MQTT cleanup is invoked.
+ *
+ * @return EXIT_SUCCESS if initialization succeeds. Otherwise EXIT_FAILURE.
+ */
+static int32_t _initDemo( void )
```
Comment (gordonwang0) [inline demos/source/aws_iot_demo_jobs.c]: This function returns EXIT_SUCCESS/EXIT_FAILURE, so its return value should be `int`.

(Also other places where EXIT_SUCCESS/EXIT_FAILURE should be int)

File: demos/source/aws_iot_demo_jobs.c [hunk 14]
```diff
@@ -0,0 +1,591 @@
...
+    return status;
+}
+
+/**
+ * @brief The callback when a new job is posted. It signals main via a semaphore.
+ */
+static void _jobsCallback( void * param1,
+                           AwsIotJobsCallbackParam_t * pCallbackInfo )
+{
+    bool success = false;
```
Comment (gordonwang0) [inline demos/source/aws_iot_demo_jobs.c]: Could you rename this `keyFound`?  More descriptive of what it's used for.

File: demos/source/aws_iot_demo_jobs.c [hunk 15]
```diff
@@ -0,0 +1,591 @@
...
+}
+
+static bool _executeCommand( const char * command,
+                             size_t commandLength,
+                             IotMqttConnection_t const mqttConnection )
+{
+#define MIN( a, b )             ( a < b ? a : b )
+#define COMPARE_COMMAND( s )    ( strncmp( command, s, MIN( commandLength, strlen( s ) ) ) == 0 )
+
+    int status = false;
```
Comment (gordonwang0) [inline demos/source/aws_iot_demo_jobs.c]: This variable should be bool.

File: demos/source/aws_iot_demo_jobs.c [hunk 16]
```diff
@@ -0,0 +1,591 @@
...
+                size_t * strLen )
+{
+    return IotJsonUtils_FindJsonValue( _jobDoc, _jobDocLength, JOB_TOPIC_KEY, JOB_TOPIC_KEY_LENGTH, str, strLen );
+}
+
+static bool _executeCommand( const char * command,
+                             size_t commandLength,
+                             IotMqttConnection_t const mqttConnection )
+{
+#define MIN( a, b )             ( a < b ? a : b )
```
Comment (gordonwang0) [inline demos/source/aws_iot_demo_jobs.c]: You don't need a min function here.  As the strncmp is always against a NULL-terminated string literal ("publish" or "print"), passing `commandLength` to strncmp is fine.

File: demos/source/aws_iot_demo_jobs.c [hunk 17]
```diff
@@ -0,0 +1,591 @@
...
+ * @param[out] msgLength the length of the topic buffer.
+ * @returns a bool. true if successful
+ */
+bool _getTopic( const char ** str,
+                size_t * strLen )
+{
+    return IotJsonUtils_FindJsonValue( _jobDoc, _jobDocLength, JOB_TOPIC_KEY, JOB_TOPIC_KEY_LENGTH, str, strLen );
+}
+
+static bool _executeCommand( const char * command,
```
Comment (gordonwang0) [inline demos/source/aws_iot_demo_jobs.c]: The JSON key is "action".  Just so the terminology is consistent, could you either rename the JSON key or replace all instances of "command" with "action"?

File: demos/source/aws_iot_demo_jobs.c [hunk 18]
```diff
@@ -0,0 +1,591 @@
...
+
+    int status = false;
+    const char * pMessage = NULL;
+    size_t messageLength = 0;
+
+    /*
+     * The "publish" command will publish a string to a specified topic.
+     *
+     * An example JSON doc:
+     * ```
```
Comment (gordonwang0) [inline demos/source/aws_iot_demo_jobs.c]: Markdown syntax will have no effect here.

File: demos/source/aws_iot_demo_jobs.c [hunk 19]
```diff
@@ -0,0 +1,591 @@
...
+    req.pThingName = pThingName;
+    req.thingNameLength = thingNameLength;
+    req.pJobId = _pJobId + 1;
+    req.jobIdLength = _JobIdLength - 2;
+    req.mallocResponse = malloc;
+
+    /* Check that a job notification came in. */
+    if( _pJobId[ 0 ] != '\0' )
+    {
+        /* Update Job to IN_PENDING. */
```
Comment (gordonwang0) [inline demos/source/aws_iot_demo_jobs.c]: Comment error: IN_PROGRESS

File: demos/source/aws_iot_demo_jobs.c [hunk 20]
```diff
@@ -0,0 +1,591 @@
...
+                                         void * pNetworkCredentialInfo,
+                                         const IotNetworkInterface_t * pNetworkInterface,
+                                         IotMqttConnection_t * const pMqttConnection )
+{
+    int32_t status = EXIT_SUCCESS;
+    IotMqttError_t connectStatus = IOT_MQTT_STATUS_PENDING;
+    IotMqttNetworkInfo_t networkInfo = IOT_MQTT_NETWORK_INFO_INITIALIZER;
+    IotMqttConnectInfo_t connectInfo = IOT_MQTT_CONNECT_INFO_INITIALIZER;
+
+    if( ( pIdentifier == NULL ) || ( pIdentifier[ 0 ] == '\0' ) )
```
Comment (gordonwang0) [inline demos/source/aws_iot_demo_jobs.c]: This check is already done in RunJobsDemo, can be removed here.

File: demos/source/aws_iot_demo_jobs.c [hunk 21]
```diff
@@ -0,0 +1,591 @@
...
+    }
+
+    /* Process the action. */
+    if( success )
+    {
+        /* Execute on the action without quotes. */
+        if( _executeCommand( pAction + 1, actionLength - 1, mqttConnection ) )
+        {
+            result = AWS_IOT_JOB_STATE_SUCCEEDED;
+        }
```
Comment (gordonwang0) [inline demos/source/aws_iot_demo_jobs.c]: If _executeCommand fails, you should still send the job update with status AWS_IOT_JOB_STATE_FAILED.

File: demos/source/aws_iot_demo_jobs.c (hunk 18 above)
Comment (nateglims) [reply demos/source/aws_iot_demo_jobs.c]: Neither does the doxygen syntax?

File: demos/source/aws_iot_demo_jobs.c (hunk 21 above)
Comment (nateglims) [reply demos/source/aws_iot_demo_jobs.c]: It is based on initialization.

File: demos/source/aws_iot_demo_jobs.c [hunk 22]
```diff
@@ -0,0 +1,618 @@
...
+    }
+
+    return success;
+
+#undef GET_ACTION
+}
+
+/**
+ * @brief The callback when a new job is posted. It signals main via a semaphore.
```
Comment (gordonwang0) [inline demos/source/aws_iot_demo_jobs.c]: Comment needs to be updated.

File: demos/source/aws_iot_demo_jobs.c [hunk 23]
```diff
@@ -0,0 +1,618 @@
...
+ */
+static void _jobsCallback( void * param1,
+                           AwsIotJobsCallbackParam_t * pCallbackInfo )
+{
+#define GET_KEY( ppStr, pLen )    IotJsonUtils_FindJsonValue( pCallbackInfo->u.callback.pDocument, pCallbackInfo->u.callback.documentLength, JOB_ID_KEY, JOB_ID_KEY_LENGTH, ppStr, pLen )
+#define GET_DOC( ppStr, pLen )    IotJsonUtils_FindJsonValue( pCallbackInfo->u.callback.pDocument, pCallbackInfo->u.callback.documentLength, JOB_DOC_KEY, JOB_DOC_KEY_LENGTH, ppStr, pLen )
+
+    bool keyFound = false;
+    bool docKeyFound = false;
+    int32_t status = EXIT_FAILURE;
```
Comment (gordonwang0) [inline demos/source/aws_iot_demo_jobs.c]: This variable isn't used.

File: demos/source/aws_iot_demo_jobs.c [hunk 24]
```diff
@@ -0,0 +1,618 @@
...
+        IotLogInfo( "New Job: %.*s", jsonValueLength, pJsonValue );
+
+        if( jsonValueLength > JOBS_DEMO_MAX_ID_LENGTH )
+        {
+            keyFound = false;
+            IotLogError( "Job ID is too long." );
+        }
+        else
+        {
+            memcpy( jobId, pJsonValue, jsonValueLength );
```
Comment (gordonwang0) [inline demos/source/aws_iot_demo_jobs.c]: No copies are needed now.  Just use the given buffer.

File: demos/source/aws_iot_demo_jobs.c [hunk 25]
```diff
@@ -0,0 +1,618 @@
...
+                                         void * pNetworkCredentialInfo,
+                                         const IotNetworkInterface_t * pNetworkInterface,
+                                         IotMqttConnection_t * const pMqttConnection )
+{
+    int32_t status = EXIT_SUCCESS;
+    IotMqttError_t connectStatus = IOT_MQTT_STATUS_PENDING;
+    IotMqttNetworkInfo_t networkInfo = IOT_MQTT_NETWORK_INFO_INITIALIZER;
+    IotMqttConnectInfo_t connectInfo = IOT_MQTT_CONNECT_INFO_INITIALIZER;
+
+    if( status == EXIT_SUCCESS )
```
Comment (gordonwang0) [inline demos/source/aws_iot_demo_jobs.c]: status is initialized to EXIT_SUCCESS, does not need to be checked here.

File: demos/source/aws_iot_demo_jobs.c [hunk 26]
```diff
@@ -0,0 +1,618 @@
...
+ * @return `EXIT_SUCCESS` if the connection is successfully established; `EXIT_FAILURE`
+ * otherwise.
+ */
+static int32_t _establishMqttConnection( const char * pIdentifier,
+                                         void * pNetworkServerInfo,
+                                         void * pNetworkCredentialInfo,
+                                         const IotNetworkInterface_t * pNetworkInterface,
+                                         IotMqttConnection_t * const pMqttConnection )
+{
+    int32_t status = EXIT_SUCCESS;
```
Comment (gordonwang0) [inline demos/source/aws_iot_demo_jobs.c]: int type for status

File: demos/source/aws_iot_demo_jobs.c [hunk 27]
```diff
@@ -0,0 +1,618 @@
...
+ * establishing the MQTT connection.
+ * @param[in] pNetworkCredentialInfo Passed to the MQTT connect function when
+ * establishing the MQTT connection.
+ * @param[in] pNetworkInterface The network interface to use for this demo.
+ * @param[out] pMqttConnection Set to the handle to the new MQTT connection.
+ *
+ * @return `EXIT_SUCCESS` if the connection is successfully established; `EXIT_FAILURE`
+ * otherwise.
+ */
+static int32_t _establishMqttConnection( const char * pIdentifier,
```
Comment (gordonwang0) [inline demos/source/aws_iot_demo_jobs.c]: int return type

File: demos/source/aws_iot_demo_jobs.c [hunk 28]
```diff
@@ -0,0 +1,618 @@
...
+        IotLogInfo( "--- Add Job using AWS CLI --- \r\n" ); /* Add an extra line for emphasis. */
+
+        for( ; ; )
+        {
+            if( Atomic_CompareAndSwap_u32( &finishFlag, 0, JOBS_DEMO_FINISHED ) == 1 )
+            {
+                break;
+            }
+
+            IotClock_SleepMs( 100 );
```
Comment (gordonwang0) [inline demos/source/aws_iot_demo_jobs.c]: This is probably too much of a busy wait.  Should probably increase to every 1 second.

File: demos/source/aws_iot_demo_jobs.c [hunk 29]
```diff
@@ -0,0 +1,618 @@
...
+        }
+    }
+
+    IotClock_SleepMs( 1000 );
```
Comment (gordonwang0) [inline demos/source/aws_iot_demo_jobs.c]: Why is this needed?

File: demos/source/aws_iot_demo_jobs.c [hunk 30]
```diff
@@ -0,0 +1,618 @@
...
+
+    if( status == EXIT_SUCCESS )
+    {
+        err = AwsIotJobs_SetNotifyNextCallback( mqttConnection, pIdentifier, thingNameLength, 0, NULL );
```
Comment (gordonwang0) [inline demos/source/aws_iot_demo_jobs.c]: This function has changed.  Please see the latest code (should not pass NULL as the last parameter).

File: demos/source/aws_iot_demo_jobs.c [hunk 31]
```diff
@@ -0,0 +1,618 @@
...
+        err = AwsIotJobs_SetNotifyNextCallback( mqttConnection, pIdentifier, thingNameLength, 0, &callbackInfo );
+
+        status = err == AWS_IOT_JOBS_SUCCESS ? EXIT_SUCCESS : EXIT_FAILURE;
+    }
+
+    if( status == EXIT_SUCCESS )
+    {
+        IotLogInfo( "--- Add Job using AWS CLI --- \r\n" ); /* Add an extra line for emphasis. */
+
+        for( ; ; )
```
Comment (gordonwang0) [inline demos/source/aws_iot_demo_jobs.c]: Why not just
```c
while( Atomic_CompareAndSwap_u32( &finishFlag, 0, JOBS_DEMO_FINISHED ) == 0 )
{
    IotClock_SleepMs(...);
}
```

File: demos/source/aws_iot_demo_jobs.c (hunk 29 above)
Comment (nateglims) [reply demos/source/aws_iot_demo_jobs.c]: It gives time for messages to be sent. I know it's not guaranteed, but you end up with a dangling "end" job if you quit right away.

File: demos/source/aws_iot_demo_jobs.c (hunk 31 above)
Comment (nateglims) [reply demos/source/aws_iot_demo_jobs.c]: This was the style used elsewhere in the code base.

File: demos/source/aws_iot_demo_jobs.c (hunk 29 above)
Comment (gordonwang0) [reply demos/source/aws_iot_demo_jobs.c]: In that case, you should set the DEMO_FINISHED flag in _updateResultCallback (the function that lets the demo know update is done).  You can use the param1 to pass context identifying the exit job.

Comment (nateglims) [reply demos/source/aws_iot_demo_jobs.c]: so casting the result to a pointer and passing it?

File: demos/source/aws_iot_demo_jobs.c (hunk 26 above)
Comment (nateglims) [reply demos/source/aws_iot_demo_jobs.c]: I've changed it, but `EXIT_SUCCESS` is a signed literal.

File: demos/src/aws_iot_demo_jobs.c [hunk 32]
```diff
@@ -0,0 +1,647 @@
...
+ * {"action":"exit"}
+ * ```
+ *
+ **/
+static bool _executeAction( const char * command,
+                            size_t commandLength,
+                            IotMqttConnection_t const mqttConnection,
+                            const char * jobDoc,
+                            size_t jobDocLength,
+                            bool * exitFlag )
```
[the remark this answers is not in the document]
Comment (nateglims) [reply demos/src/aws_iot_demo_jobs.c]: The return value is the success status of executing the action. The `exitFlag` is a flag for if the program should exit.

File: demos/src/aws_iot_demo_jobs.c [hunk 33]
```diff
@@ -0,0 +1,647 @@
...
+/**
+ * @brief Execute the Jobs Demo.
+ *
+ * This is the top level function called by the IOT Demo harness.
+ */
+int RunJobsDemo( bool awsIotMqttMode,
+                 const char * pIdentifier,
+                 void * pNetworKServerInfo,
+                 void * pNetworkCredentialInfo,
+                 const IotNetworkInterface_t * pNetworkInterface );
```
[the remark this answers is not in the document]
Comment (nateglims) [reply demos/src/aws_iot_demo_jobs.c]: I was following the shadow demo for consistency. Not sure why that one has a forward declaration.

File: demos/src/aws_iot_demo_jobs.c [hunk 34]
```diff
@@ -0,0 +1,587 @@
...
+/* JSON Utilies for pulling out values. */
+#include "iot_json_utils.h"
+
+/* Jobs Library. */
+#include "aws_iot_jobs.h"
+
+/**
+ * @brief The Library Timeout.
+ */
+#define TIMEOUT_MS                      ( 5000u )
```
[the remark this answers is not in the document]
Comment (nateglims) [reply demos/src/aws_iot_demo_jobs.c]: We should add this to the CSDK Developer guide.
